index.rst

```diff
@@ -1,61 +1,486 @@
+- a record of which predicted inputs were ``actually_used``.
+
+These can easily be saved to a file (e.g. JSON) written by the quantum-execution harness, and here the design ties again into the quantum-backed butler concept, which also needs to write per-quantum files in order to save datastore record data.
+Just like the provenance we wish to save here, the eventual home of those datastore records is the shared ``Registry`` database, so it is extremely natural to save them both to the same files, and upload provenance when the datasets themselves are ingested in bulk after execution completes.
+
+In fact, these per-quantum files may also help solve yet another problem; as described in `DMTN-213 <https://dmtn-213.lsst.io/>`_, our approach to multi-site processing and synchronization will probably involve metadata files that are transferred along with dataset files by Rucio, in order to ensure enough information for butler ingest is available from files alone.
+These provenance files could easily play that role as well.
+
+The quantum-backed butler design is a solution to a problem unique to at-scale batch processing, so writing the ``QuantumGraph`` and provenance files to BPS-managed locations (such as its "submit" directory) there is completely fine.
+That's not true for provenance, which we want to work regardless of how execution is performed.
+This is related to the long-running middleware goal of better integrating ``pipetask`` and BPS.
+It will probably also involve carving out a third aspect of butler (a new sibling to ``Registry`` and ``Datastore``) for ``QuantumGraph`` and provenance files, because
+
+- like ``Datastore``, this aspect would be backed by files in a shared filesystem or object store (but not necessarily the same filesystem or bucket as an associated ``Datastore``);
+- like ``Registry``, the new aspect would provide descriptive and organizational metadata for datasets, rather than hold datasets themselves, and after execution its content would be completely loaded into the ``Registry``.
+
+The full high-level design will be the subject of a future technote, but an early sketch can be found `in Confluence <https://confluence.lsstcorp.org/display/DM/Saving+per-Quantum+provenance+and+propagating+nothing-to-do+cases%2C+and+The+Future>`_.
+
+Recording provenance only when using BPS (and relying on it to manage the ``QuantumGraph`` and provenance files) in the interim seems like a good first step.
+Extending the design to include non-BPS processing may take time, but we do not anticipate it changing what happens at a low level or the appearance of persisted provenance information.
+
+.. _querying-provenance:
+
+Interfaces for querying quantum provenance
+------------------------------------------
+
+Given the similarity between quanta and datasets in terms of table structure, a ``Registy.queryQuanta`` method analogous to ``queryDatasets`` provides a good starting point for provenance searches::
```

Comment:
Do we have requirements on provenance query API/functionality? I'm worried that trying to provide a very generic API we make things harder for ourselves.

Reply:
We effectively have a global requirement to support the IVOA ProvDAL standard (which effectively means a way of implementing the ProvDM standard), although the server supporting such a thing vs butler registry supporting it natively can be two separate things.